File: objectiveai-py/objectiveai/functions/executions/http.py
"""HTTP functions for function executions."""

from __future__ import annotations

from typing import TYPE_CHECKING, Union

if TYPE_CHECKING:
    from objectiveai.client import ObjectiveAI
    from objectiveai.functions.executions.request import FunctionExecutionCreateParams
    from objectiveai.functions.executions.response.streaming import FunctionExecutionChunk
    from objectiveai.functions.executions.response.unary import FunctionExecution
    from objectiveai.stream import Stream


async def create_function_execution(
    client: ObjectiveAI,
    params: FunctionExecutionCreateParams,
    *,
    fremote: str | None = None,
    fowner: str | None = None,
    frepository: str | None = None,
    fcommit: str | None = None,
    premote: str | None = None,
    powner: str | None = None,
    prepository: str | None = None,
    pcommit: str | None = None,
) -> Union[FunctionExecution, Stream[FunctionExecutionChunk]]:
    """Execute a function.

    Path is built from the remote source arguments. If none are provided,
    uses the inline execution path. If ``params.stream`` is true, returns
    a streaming response.
    """
    if fremote and fowner and frepository:
        parts = [f"functions/{fremote}/{fowner}/{frepository}"]
        if fcommit:
            parts.append(fcommit)
        if premote and powner and prepository:
            parts.append(f"profiles/{premote}/{powner}/{prepository}")
            if pcommit:
                parts.append(pcommit)
        path = "/".join(parts)
    else:
        path = "functions"

    if getattr(params, "stream", None):
        return await client.post_streaming(path, params)
    return await client.post_unary(path, params)

File: objectiveai-js/src/functions/executions/http.ts
import { ObjectiveAI, type RequestOptions } from "../../client";
import { Stream } from "../../stream";
import type { FunctionsExecutionsRequestRequest } from "./request/request";
import type { FunctionsExecutionsResponseUnaryFunctionExecution } from "./response/unary/functionExecution";
import type { FunctionsExecutionsResponseStreamingFunctionExecutionChunk } from "./response/streaming/functionExecutionChunk";

function buildExecutionPath(request: FunctionsExecutionsRequestRequest): string {
  if (!("path" in request)) {
    return "/functions";
  }
  const { path } = request;
  if ("fremote" in path && "premote" in path) {
    let url = `/functions/${path.fremote}/${path.fowner}/${path.frepository}`;
    if (path.fcommit != null) url += `/${path.fcommit}`;
    url += `/profiles/${path.premote}/${path.powner}/${path.prepository}`;
    if (path.pcommit != null) url += `/${path.pcommit}`;
    return url;
  }
  if ("fremote" in path) {
    let url = `/functions/${path.fremote}/${path.fowner}/${path.frepository}`;
    if (path.fcommit != null) url += `/${path.fcommit}`;
    return url;
  }
  let url = `/functions/profiles/${path.premote}/${path.powner}/${path.prepository}`;
  if (path.pcommit != null) url += `/${path.pcommit}`;
  return url;
}

export function functionsExecutionsCreateFunctionExecution(
  client: ObjectiveAI,
  request: FunctionsExecutionsRequestRequest & { body: { stream: true } },
  options?: RequestOptions,
): Promise<Stream<FunctionsExecutionsResponseStreamingFunctionExecutionChunk>>;
export function functionsExecutionsCreateFunctionExecution(
  client: ObjectiveAI,
  request: FunctionsExecutionsRequestRequest & { body: { stream?: false | null } },
  options?: RequestOptions,
): Promise<FunctionsExecutionsResponseUnaryFunctionExecution>;
export function functionsExecutionsCreateFunctionExecution(
  client: ObjectiveAI,
  request: FunctionsExecutionsRequestRequest,
  options?: RequestOptions,
): Promise<
  | Stream<FunctionsExecutionsResponseStreamingFunctionExecutionChunk>
  | FunctionsExecutionsResponseUnaryFunctionExecution
> {
  const path = buildExecutionPath(request);
  if (request.body.stream) {
    return client.post_streaming<FunctionsExecutionsResponseStreamingFunctionExecutionChunk>(
      path,
      request.body,
      options,
    );
  }
  return client.post_unary<FunctionsExecutionsResponseUnaryFunctionExecution>(
    path,
    request.body,
    options,
  );
}
